File: android/core/network/src/main/kotlin/xyz/klinik/network/PatientsApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class Patient(
    val id: String,
    /** Human-facing file number, e.g. 2026-K7RMPX. */
    val mrn: String,
    val firstName: String,
    val lastName: String,
    val birthDate: String,
    val sex: String,
    val country: String,
    val city: String? = null,
    val preferredLanguage: String,
    val status: String,
    val createdAt: String,
) {
    val fullName: String get() = "$firstName $lastName"
}

@Serializable
data class PatientPage(
    val items: List<Patient>,
    /** Null on the last page. */
    val nextCursor: String? = null,
)

data class PatientSearch(
    val query: String? = null,
    val country: String? = null,
    val status: String? = null,
    val cursor: String? = null,
    val limit: Int? = null,
) {
    fun toQuery(): Map<String, String> = buildMap {
        query?.takeIf { it.isNotEmpty() }?.let { put("q", it) }
        country?.let { put("country", it) }
        status?.let { put("status", it) }
        cursor?.let { put("cursor", it) }
        limit?.let { put("limit", it.toString()) }
    }
}

/**
 * A new file, with the five things a record cannot be opened without.
 *
 * Country is required because it decides the language the patient is written
 * to in and which discharge advice they get — a file with no country is a file
 * the clinic has to guess about.
 */
@Serializable
data class NewPatient(
    val firstName: String,
    val lastName: String,
    /** `1984-03-21`. */
    val birthDate: String,
    val sex: String,
    /** ISO 3166-1 alpha-2. */
    val country: String,
    val city: String? = null,
    val nationality: String? = null,
    val preferredLanguage: String? = null,
    /** Instagram, Google, an agency, a referral — for the channel report. */
    val referralSource: String? = null,
    val assignedDoctorId: String? = null,
    val agencyId: String? = null,
)

class PatientsApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    suspend fun search(search: PatientSearch): PatientPage =
        decode(client.send(Endpoint(HttpMethod.GET, "patients", query = search.toQuery())))

    suspend fun detail(id: String): Patient =
        decode(client.send(Endpoint(HttpMethod.GET, "patients/$id")))

    /** Opens a file. The server assigns the file number. */
    suspend fun create(patient: NewPatient): Patient =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.POST,
                    "patients",
                    body = json.encodeToString(NewPatient.serializer(), patient),
                ),
            ),
        )

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}

// MARK: - Patient-facing

@Serializable
data class NextAppointment(
    val id: String,
    val scheduledAt: String,
    val type: String,
    val location: String? = null,
)

@Serializable
data class HomePatient(
    val id: String,
    val mrn: String,
    val firstName: String,
    val lastName: String,
    val preferredLanguage: String,
    val status: String,
) {
    val fullName: String get() = "$firstName $lastName"
}

File: ios/Sources/KlinikAPI/PatientsAPI.swift
import Foundation
import KlinikCore

public struct Patient: Decodable, Sendable, Identifiable, Equatable {
    public let id: String
    /// Human-facing file number, e.g. 2026-K7RMPX.
    public let mrn: String
    public let firstName: String
    public let lastName: String
    public let birthDate: Date
    public let sex: String
    public let country: String
    public let city: String?
    public let preferredLanguage: String
    public let status: String
    public let createdAt: Date

    public var fullName: String { "\(firstName) \(lastName)" }
}

public struct PatientPage: Decodable, Sendable {
    public let items: [Patient]
    /// Null on the last page.
    public let nextCursor: String?
}

public struct PatientSearch: Sendable, Equatable {
    public var query: String?
    public var country: String?
    public var status: String?
    public var cursor: String?
    public var limit: Int?

    public init(
        query: String? = nil,
        country: String? = nil,
        status: String? = nil,
        cursor: String? = nil,
        limit: Int? = nil
    ) {
        self.query = query
        self.country = country
        self.status = status
        self.cursor = cursor
        self.limit = limit
    }

    var queryItems: [String: String] {
        var items: [String: String] = [:]

        if let query, !query.isEmpty { items["q"] = query }
        if let country { items["country"] = country }
        if let status { items["status"] = status }
        if let cursor { items["cursor"] = cursor }
        if let limit { items["limit"] = String(limit) }

        return items
    }
}

/**
 * A new patient file (spec M2).
 *
 * The file number is not here: the server allocates it. A client that could
 * choose one could collide with an existing file, and the number is what a
 * clinic writes on paper.
 */
public struct NewPatient: Encodable, Sendable {
    public let firstName: String
    public let lastName: String
    public let birthDate: Date
    /// `FEMALE`, `MALE` or `OTHER` — the server's enum.
    public let sex: String
    /// ISO 3166-1 alpha-2. Drives the language and the discharge advice.
    public let country: String
    public let city: String?
    public let nationality: String?
    public let preferredLanguage: String?
    public let referralSource: String?
    public let assignedDoctorId: String?

    public init(
        firstName: String,
        lastName: String,
        birthDate: Date,
        sex: String,
        country: String,
        city: String? = nil,
        nationality: String? = nil,
        preferredLanguage: String? = nil,
        referralSource: String? = nil,
        assignedDoctorId: String? = nil
    ) {
        self.firstName = firstName
        self.lastName = lastName
        self.birthDate = birthDate
        self.sex = sex
        self.country = country
        self.city = city
        self.nationality = nationality
        self.preferredLanguage = preferredLanguage
        self.referralSource = referralSource
        self.assignedDoctorId = assignedDoctorId
    }
}

public struct PatientsAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    /// Opens a file. The server allocates the file number and returns it.
    public func create(_ patient: NewPatient) async throws -> Patient {
        try await client.send(
            Endpoint(
                method: .post,
                path: "patients",
                body: try JSONEncoder.klinik.encode(patient)
            ),
            as: Patient.self
        )
    }

    public func search(_ search: PatientSearch) async throws -> PatientPage {
        try await client.send(
            Endpoint(method: .get, path: "patients", query: search.queryItems),
            as: PatientPage.self
        )
    }

    public func detail(id: String) async throws -> Patient {
        try await client.send(Endpoint(method: .get, path: "patients/\(id)"), as: Patient.self)
    }
}

// MARK: - Patient-facing

public struct NextAppointment: Decodable, Sendable, Equatable {
    public let id: String
    public let scheduledAt: Date
    public let type: String
    public let location: String?
}

/// Everything the patient home screen needs, in one call.
public struct PatientHomeSummary: Decodable, Sendable, Equatable {
    public struct HomePatient: Decodable, Sendable, Equatable {
        public let id: String
        public let mrn: String
        public let firstName: String
        public let lastName: String
        public let preferredLanguage: String
        public let status: String

        public var fullName: String { "\(firstName) \(lastName)" }
    }

    public let patient: HomePatient
    public let nextAppointment: NextAppointment?
    /// Doses scheduled for today that are still waiting.
    public let medicationsDueToday: Int
    public let unreadMessages: Int
    /// Mandatory pre-op documents not yet uploaded (spec M17).
    public let missingDocuments: Int
}

public struct MeAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    /// So an extension in another file can reach the client without exposing it.
    func send<T: Decodable & Sendable>(_ endpoint: Endpoint, as type: T.Type) async throws -> T {
        try await client.send(endpoint, as: type)
    }

    public func summary() async throws -> PatientHomeSummary {
        try await client.send(
            Endpoint(method: .get, path: "me/summary"),
            as: PatientHomeSummary.self
        )
    }
}
